File: _inc/laravel/app/Http/Controllers/Info/NotificationTemplatesController.php
<?php

namespace App\Http\Controllers\Info;

use App\Http\Controllers\Abstracts\Controller;
use function App\Http\Controllers\Helpers\{defaultUndefinedException, defaultPermissionDenial};
use App\Config\Constants\{
    DatabaseConstants,
    MiddlewaresConstants,
    ViewsConstants
};
use App\Models\{
    Language,
    NotificationTemplateLang,
    NotificationTemplate,
    Utility
};
use App\Traits\ChecksLogin;
use App\Traits\ChecksPermissions;
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\Http\{
    RedirectResponse,
    Request
};
use Illuminate\Support\Facades\{
    Log,
    Validator,
    View as ViewFacade
};
use Illuminate\View\View;
use Symfony\Component\HttpFoundation\Response;

use function App\Http\Controllers\Helpers\{defaultUndefinedException, defaultPermissionDenial};
use App\Traits\DefinesResourceActions;
class NotificationTemplateController extends Controller
{
	use DefinesResourceActions;

    use ChecksLogin, ChecksPermissions;

    public function __construct()
    {
        $this->middleware([MiddlewaresConstants::XSS]);
    }

    public function index(Request $request, int|string $id, string $lang = DatabaseConstants::DEFAULT_LANG): View|Response
    {
        $cls    = __CLASS__;
        $action = __FUNCTION__;
        $view   = ViewsConstants::NTF_TMP . '.index';

        return $this->measureProfile("$cls::$action", function () use ($request, $id, $lang, $cls, $action, $view) {
            try {
                if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
                if (!($denial = $this->guard($request, 'manage notification template'))) return $denial;
                $template = $id ? NotificationTemplate::find($id) : NotificationTemplate::first();
                if (!$template) return redirect()->back()->with('error', __('Not exists in notification template.'));
                $languages  = Utility::languages();
                $langName   = Language::where('code', $lang)->first();
                $translation = NotificationTemplateLang::where('parent_id', $template->id)
                    ->where('lang', $lang)
                    ->where(DatabaseConstants::COL_TABLE_CREATOR, $request->user()->creatorId())
                    ->first()
                    ?: NotificationTemplateLang::where('parent_id', $template->id)
                    ->where('lang', $lang)
                    ->first()
                    ?: tap(
                        NotificationTemplateLang::where('parent_id', $template->id)
                            ->where('lang', DatabaseConstants::DEFAULT_LANG)
                            ->first(),
                        function ($t) use ($lang) {
                            if ($t) $t->lang = $lang;
                        }
                    );

                $allTemplates = NotificationTemplate::all();

                if (!ViewFacade::exists($view))
                    return defaultUndefinedException($request, new \RuntimeException('View not found'), "$cls::$action");

                return view($view, [
                    'notificationTemplate'  => $template,
                    'notificationTemplates' => $allTemplates,
                    'currentTemplateLang'   => $translation,
                    DatabaseConstants::TABLE_LANGS => $languages,
                    'langName'              => $langName,
                ]);
            } catch (AuthorizationException $e) {
                return defaultPermissionDenial($request, $e, "$cls::$action");
            } catch (\Throwable $e) {
                return defaultUndefinedException($request, $e, "$cls::$action");
            }
        });
    }

    /**
     * Update or create a translation for a notification template.
     */
    public function update(Request $request, int|string $id): RedirectResponse
    {
        $cls    = __CLASS__;
        $action = __FUNCTION__;
        $route  = ViewsConstants::NTF_TMP . '.index';

        return $this->measureProfile("$cls::$action", function () use ($request, $id, $cls, $action, $route) {
            try {
                if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
                if (!($resp = $this->guard($request, 'edit notification template'))) return $resp;

                $v = Validator::make($request->all(), ['content' => 'required']);
                if ($v->fails()) return redirect()->back()->with('error', $v->errors()->first());

                $lang      = $request->input('lang');
                $content   = $request->input('content');
                $creatorId = $request->user()->creatorId();

                $record = NotificationTemplateLang::where('parent_id', $id)
                    ->where('lang', $lang)
                    ->where(DatabaseConstants::COL_TABLE_CREATOR, $creatorId)
                    ->first();

                if (!$record) {
                    $variables = NotificationTemplateLang::where('parent_id', $id)
                        ->where('lang', $lang)
                        ->first()
                        ?->variables;

                    $record = new NotificationTemplateLang();
                    $record->parent_id  = $id;
                    $record->lang       = $lang;
                    $record->variables  = $variables;
                    $record->created_by = $creatorId;
                }

                $record->content = $content;
                $record->save();
                Log::info('Notification template lang saved', ['id' => $record->id]);

                return redirect()->route($route, [$id, $lang])
                    ->with('success', __('Notification Template successfully updated.'));
            } catch (AuthorizationException $e) {
                return defaultPermissionDenial($request, $e, "$cls::$action");
            } catch (\Throwable $e) {
                return defaultUndefinedException($request, $e, "$cls::$action");
            }
        });
    }
}

File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/info/notification_templates_controller.py
import inspect
from typing import Union
from django.core.exceptions import PermissionDenied
from django.contrib import messages
from django.http import HttpRequest, HttpResponse
from django.shortcuts import render, get_object_or_404, redirect
from .._helpers.http import get_redirect_url
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception
from . import Controller
from ....Models.info.notification_templates import NotificationTemplates
from ....Models.info.notification_template_lang import NotificationTemplateLangs
from ....Models.utils.utility import Utility
from ....Models.shapes.language import Language

class NotificationTemplatesController(Controller):
  
  @classmethod
  def _set_auth(cls, request: HttpRequest, perm:str) -> Union[bool, Exception]:
      self = cls()
      self.request = request 
      self.authorize(f'{perm} notification templates')

  @classmethod
  def index(cls, request: HttpRequest, id=None, lang='en') -> HttpResponse:
    REF=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    try:
      NotificationTemplatesController._set_auth(request, 'view')
      template = NotificationTemplates.objects.filter(pk=id).first() if id else NotificationTemplates.objects.first()
      if not template:
        messages.error(request,'Template does not exist.') 
        return redirect(get_redirect_url(request))
      langs = Utility.languages()
      lang_obj = Language.objects.filter(code=lang).first()
      noti_lang = NotificationTemplateLangs.objects.filter(parent_id=template.id,lang=lang,created_by=request.user.creator_id()).first() \
                 or NotificationTemplateLangs.objects.filter(parent_id=template.id,lang=lang).first() \
                 or NotificationTemplateLangs.objects.filter(parent_id=template.id,lang='en').first()
      all_templates = NotificationTemplates.objects.all()
      return render(request,'notification_templates/index.html',{
        'notification_template': template,
        'notification_templates': all_templates,
        'curr_noti_tempLang': noti_lang,
        'languages': langs,
        'LangName': lang_obj
      })
    except PermissionDenied as err:
      return default_permission_denial(request,err,ref=REF)
    except Exception as err:
      return default_undefined_exception(request,err,ref=REF)

  @classmethod
  def create(cls, request: HttpRequest) -> HttpResponse:
    REF=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    try:
      NotificationTemplatesController._set_auth(request, 'create')
      langs = Utility.languages()
      return render(request,'notification_templates/create.html',{
        'languages': langs
      })
    except PermissionDenied as err:
      return default_permission_denial(request,err,ref=REF)
    except Exception as err:
      return default_undefined_exception(request,err,ref=REF)

  @classmethod
  def store(cls, request: HttpRequest) -> HttpResponse:
    REF=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    try:
      NotificationTemplatesController._set_auth(request, 'create')
      data = request.POST
      name = data.get('name')
      variables = data.get('variables')
      if not name or not variables:
        messages.error(request,'Name and variables are required.') 
        return redirect(get_redirect_url(request))
      NotificationTemplates.objects.create(
        name=name,variables=variables,created_by=request.user.creator_id()
      )
      messages.success(request,'Template successfully created.')
      return redirect(get_redirect_url(request))
    except PermissionDenied as err:
      return default_permission_denial(request,err,ref=REF)
    except Exception as err:
      return default_undefined_exception(request,err,ref=REF)

  @classmethod
  def show(cls, request: HttpRequest, id) -> HttpResponse:
    REF=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    try:
      NotificationTemplatesController._set_auth(request, 'view')
      tpl = get_object_or_404(NotificationTemplates,pk=id)
      langs = Utility.languages()
      return render(request,'notification_templates/show.html',{
        'notification_template': tpl,
        'languages': langs
      })
    except PermissionDenied as err:
      return default_permission_denial(request,err,ref=REF)
    except Exception as err:
      return default_undefined_exception(request,err,ref=REF)

  @classmethod
  def edit(cls, request: HttpRequest, id) -> HttpResponse:
    REF=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    try:
      NotificationTemplatesController._set_auth(request, 'update')
      tpl = get_object_or_404(NotificationTemplates,pk=id)
      langs = Utility.languages()
      return render(request,'notification_templates/edit.html',{
        'notification_template': tpl,
        'languages': langs
      })
    except PermissionDenied as err:
      return default_permission_denial(request,err,ref=REF)
    except Exception as err:
      return default_undefined_exception(request,err,ref=REF)

  @classmethod
  def destroy(cls, request: HttpRequest, id) -> HttpResponse:
    REF=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    try:
      NotificationTemplatesController._set_auth(request, 'delete')
      tpl = get_object_or_404(NotificationTemplates,pk=id)
      tpl.delete()
      messages.success(request,'Template successfully deleted.')
      return redirect(get_redirect_url(request))
    except PermissionDenied as err:
      return default_permission_denial(request,err,ref=REF)
    except Exception as err:
      return default_undefined_exception(request,err,ref=REF)
